Ranking IK Solutions

aditya-vk opened the pull request:
This PR introduces ranker for IK solutions.

Looking for discussion on:
1. Should we make it more general to also be able to rank TSRs with the same (templated) classes? 
I prefer having them separate to make it simple and free of potential errors.
  - Decided to keep them separate
2. Where do we have the files? It seemed `constraint/dart/` might be a good place since it's still a constraint over the IKs which by itself is a constraint. Can also be under `planner`.
  - Decided to have them under `distance`
3. Naming convention for the classes.
4. Efficiency and implementation details.

~WIP: Tests. I started a test for FIFO (which is more or less a random IK ranker like in `prpy`).~
***

**Before creating a pull request**

- [x] Document new methods and classes
- [x] Format code with `make format`

**Before merging a pull request**

- [x] Set version target by selecting a milestone on the right side
- [x] Summarize this change in `CHANGELOG.md`
- [x] Add unit test(s) for this change

brianhou commented on include/aikido/constraint/dart/IKRankingStrategy.hpp:
```diff
@@ -0,0 +1,69 @@
+  virtual ~IKRankingStrategy() = default;
+
+  /// Returns the statespace.
+  statespace::ConstStateSpacePtr getStateSpace() const;
```
Is this necessary? What do we think we would need the state space for?

brianhou replied:
Oh, is this for the `JointLimitAvoidanceStrategy`?

aditya-vk replied:
That was the idea, yeah.

brianhou commented on include/aikido/constraint/dart/IKRankingStrategy.hpp:
```diff
@@ -0,0 +1,69 @@
+
+  /// Add IK solution to Ranker
+  /// Evaluates, creates a pair and stores it in the vector.
+  void addIKSolution(statespace::StateSpace::State* solution);
```
I think it might be easier to require that we're given a vector of states, rather than adding one state at a time.

aditya-vk replied:
Moreover, since there is no notion of removing IKs from the container, what you suggest is much neater than having this function, true! Good point!

brianhou commented on include/aikido/constraint/dart/IKRankingStrategy.hpp:
```diff
@@ -0,0 +1,69 @@
+  /// \param[in] numIKSolution Number of IK solutions to rank.
+  IKRankingStrategy(
+      statespace::dart::ConstMetaSkeletonStateSpacePtr metaSkeletonStateSpace,
+      ::dart::dynamics::ConstMetaSkeletonPtr metaSkeleton,
```
What would the `MetaSkeleton` be used for? If it's just to extract the current configuration, maybe this should be `currentState` instead?

brianhou commented:
> Should we make it more general to also be able to rank TSRs with the same (templated) classes? I prefer having them separate to make it simple and free of potential errors.

Separate sounds fine for me. If we implement both and find that they're more similar than we thought, maybe we can merge them then.

> Where do we have the files? It seemed `constraint/dart/` might be a good place since it's still a constraint over the IKs which by itself is a constraint. Can also be under `planner`.

I don't really consider this ranking a constraint. Maybe `aikido::distance`?

aditya-vk commented on include/aikido/distance/RandomConfigurationRanker.hpp:
```diff
@@ -0,0 +1,30 @@
+namespace aikido {
+namespace distance {
+
+class RandomConfigurationRanker : public ConfigurationRanker
```
Is this class required, since the configurations are anyway generated in a random order (except maybe when sequential sampleable with start state is used).

brianhou replied:
I don't think this is necessary, unless we decide that we want to have a "default" ranker (and we don't want that defualt to be something like `NominalConfigurationRanker`).

brianhou commented on CHANGELOG.md:
```diff
@@ -8,6 +8,10 @@
 
+* Distance
+
+	* Added methods to rank configurations based on specified metric: [#423](https://github.com/personalrobotics/aikido/pull/423)
```
Nit: indent with spaces, not tabs.

brianhou commented on include/aikido/distance/ConfigurationRanker.hpp:
```diff
@@ -0,0 +1,52 @@
+  ConfigurationRanker(
+      statespace::dart::ConstMetaSkeletonStateSpacePtr metaSkeletonStateSpace,
+      ::dart::dynamics::ConstMetaSkeletonPtr metaSkeleton,
+      const std::vector<statespace::StateSpace::State*> ikSolutions);
```
I don't think we need to take in the `ikSolutions` here. Maybe we'd want to reuse the ranker with a different set of `ikSolutions`?

brianhou commented on include/aikido/distance/ConfigurationRanker.hpp:
```diff
@@ -0,0 +1,52 @@
+  statespace::ConstStateSpacePtr getStateSpace() const;
+
+  /// Returns the vector of ranked IK solutions.
+  std::vector<std::pair<statespace::StateSpace::State*, double>>&
```
I think this method might be more useful if it's something like `std::vector<State*>& rank(std::vector<State*>& ikSolutions)`. Perhaps it can take in the vector of states and sort them in place? I don't think it's necessary for an external user to see the resulting scores.

brianhou commented on include/aikido/distance/ConfigurationRanker.hpp:
```diff
@@ -0,0 +1,52 @@
+  ::dart::dynamics::ConstMetaSkeletonPtr mMetaSkeleton;
+
+  /// Vector to hold IK solutions and corresponding score.
+  std::vector<std::pair<statespace::StateSpace::State*, double>> mIKSolutions;
```
I don't think this is necessary.

brianhou commented on src/distance/ConfigurationRanker.cpp:
```diff
@@ -0,0 +1,64 @@
+    throw std::invalid_argument("Vector of IK Solutions is empty.");
+
+  mIKSolutions.resize(ikSolutions.size());
+  mIKSolutions.shrink_to_fit();
```
I think not having `mIKSolutions` allows us to skip all this initialization code.

brianhou commented on src/distance/JointAvoidanceConfigurationRanker.cpp:
```diff
@@ -0,0 +1,60 @@
+  mMetaSkeletonStateSpace->convertPositionsToState(
+      lowerLimitPosition, upperLimitState);
+
+  auto dmetric = createDistanceMetricFor(
```
Should this distance metric be a member variable in `ConfigurationRanker` that we can reuse?

brianhou commented on src/distance/NominalConfigurationRanker.cpp:
```diff
@@ -0,0 +1,36 @@
+double NominalConfigurationRanker::evaluateIKSolution(
+    statespace::StateSpace::State* solution) const
+{
+  auto currentState = mMetaSkeletonStateSpace->createState();
```
Should this just be `mMetaSkeletonStateSpace->getScopedStateFromMetaSkeleton(mMetaSkeleton.get())`?

brianhou commented on include/aikido/distance/ConfigurationRanker.hpp:
```diff
@@ -0,0 +1,54 @@
+  /// Returns the statespace.
+  statespace::ConstStateSpacePtr getStateSpace() const;
+
+  /// Returns the vector of ranked configurations.
```
Nit: update this docstring now that we're modifying the vector rather than returning it?

brianhou commented on include/aikido/distance/JointAvoidanceConfigurationRanker.hpp:
```diff
@@ -0,0 +1,32 @@
+  double evaluateConfiguration(
+      statespace::StateSpace::State* solution) const override;
+
+  std::vector<std::size_t> mUnboundedLowerLimitsIndices;
```
Nit: document these member variables?

brianhou commented on include/aikido/distance/NominalConfigurationRanker.hpp:
```diff
@@ -0,0 +1,34 @@
+  double evaluateConfiguration(
+      statespace::StateSpace::State* solution) const override;
+
+  const statespace::StateSpace::State* mNominalConfiguration;
```
Nit: document this.

brianhou commented on src/distance/ConfigurationRanker.cpp:
```diff
@@ -0,0 +1,71 @@
+    scoredConfigurations[i].second = evaluateConfiguration(configurations[i]);
+  }
+
+  struct sortingFunction
```
Can we move this out of this method into an anonymous namespace? Also, maybe a more descriptive name would be helpful.

brianhou commented on src/distance/ConfigurationRanker.cpp:
```diff
@@ -0,0 +1,71 @@
+      scoredConfigurations.begin(),
+      scoredConfigurations.end(),
+      std::back_inserter(configurations),
+      [](const std::pair<statespace::CartesianProduct::State*, int>& item) {
```
Is `int` correct? Seems like it should be `double`.

brianhou commented on src/distance/JointAvoidanceConfigurationRanker.cpp:
```diff
@@ -0,0 +1,56 @@
+  for (auto index : mUnboundedUpperLimitsIndices)
+    upperLimits[index] = solutionPosition[index];
+
+  auto lowerLimitsState
```
It seems that we're going to be creating the `lowerLimitsState` and `upperLimitsState` each time we call `evaluateConfiguration`. Can the initialization of those states go in the constructor instead? Then we can call `convertPositionsToState` with the updated position vector.

brianhou commented on include/aikido/distance/ConfigurationRanker.hpp:
```diff
@@ -0,0 +1,54 @@
+      std::vector<statespace::CartesianProduct::State*>& configurations);
+
+protected:
+  /// Returns the score of the configuration
```
I think we should document that a lower "score" means that it'll be ranked first? Score then seems like a confusing term for it, since we generally want high scores -- maybe "cost"?

brianhou commented on include/aikido/statespace/dart/MetaSkeletonStateSpace.hpp:
```diff
@@ -171,6 +171,12 @@ class MetaSkeletonStateSpace : public CartesianProduct
+  /// specified DART \c MetaSkeleton positions.
+  ///
+  /// \param[in] positions DART \c MetaSkeleton positions
+  ScopedState getStateFromPositions(Eigen::VectorXd& positions) const;
```
I'm not sure whether this method is useful.

aditya-vk replied:
I see the following code block repeating a lot:
```
auto state = space->createState();
space->convertPositionsToState(position, state);
```
Figured I'd make a helper function.

brianhou replied:
In that case, let's add it as a separate/future PR instead so this one is just about the IK rankers.

brianhou commented on tests/distance/test_JointAvoidanceConfigurationRanker.cpp:
```diff
@@ -0,0 +1,127 @@
+
+  Eigen::VectorXd rankedStateOne(2), rankedStateTwo(2), rankedStateThree(2);
+  mStateSpace->convertStateToPositions(
+      mStateSpace->cloneState(states[0]), rankedStateOne);
```
Why are these `cloneState`s necessary?

brianhou commented on tests/distance/test_JointAvoidanceConfigurationRanker.cpp:
```diff
@@ -0,0 +1,127 @@
+{
+  Eigen::VectorXd jointPosition(2);
+
+  jointPosition << 0.3, 0.1;
```
There's a lot of duplicated code here, can we try to put this into a loop? We're creating the `Eigen::Vector2d` objects below anyway, so maybe we can use them up here.

brianhou commented on tests/distance/test_JointAvoidanceConfigurationRanker.cpp:
```diff
@@ -0,0 +1,127 @@
+  mStateSpace->convertStateToPositions(
+      mStateSpace->cloneState(states[2]), rankedStateThree);
+
+  EXPECT_TRUE(rankedStateOne.isApprox(Eigen::Vector2d(0.3, 0.1)));
```
Why wouldn't these be identical? Also, I think we should use `EXPECT_EIGEN_EQUAL` here.

aditya-vk commented:
Cleaned up! Should be good to go (I think). Will follow up with updates to planToTSR methods once this gets merged :)

The tests are failing for reasons that don't have to do with the code itself.

brianhou commented on include/aikido/distance/ConfigurationRanker.hpp:
```diff
@@ -0,0 +1,56 @@
+#ifndef AIKIDO_DISTANCE_CONFIGURATIONRANKER_HPP_
+#define AIKIDO_DISTANCE_CONFIGURATIONRANKER_HPP_
+
+#include "aikido/distance/CartesianProductWeighted.hpp"
```
Nit: is this include necessary?

brianhou commented on include/aikido/distance/ConfigurationRanker.hpp:
```diff
@@ -0,0 +1,56 @@
+
+  /// Returns a vector of configurations ranked in increasing order of costs.
+  /// \param[in] configurations Vector of configurations to rank.
+  std::vector<statespace::CartesianProduct::State*> rankConfigurations(
```
Let's change this from `CartesianProduct::State` to `MetaSkeletonStateSpace::State`, etc.

brianhou commented on include/aikido/distance/ConfigurationRanker.hpp:
```diff
@@ -0,0 +1,56 @@
+#include "aikido/distance/CartesianProductWeighted.hpp"
+#include "aikido/distance/DistanceMetric.hpp"
+#include "aikido/distance/defaults.hpp"
+#include "aikido/statespace/CartesianProduct.hpp"
```
I think this is also unnecessary.

brianhou commented on include/aikido/distance/ConfigurationRanker.hpp:
```diff
@@ -0,0 +1,56 @@
+  virtual ~ConfigurationRanker() = default;
+
+  /// Returns the statespace.
+  statespace::ConstStateSpacePtr getStateSpace() const;
```
This method doesn't seem to be used. I'd prefer to remove it if we don't see a use for it, but I think this should return a `ConstMetaSkeletonStateSpacePtr` if we decide to keep it.

aditya-vk replied:
I'll remove it for now.

brianhou commented on include/aikido/distance/NominalConfigurationRanker.hpp:
```diff
@@ -0,0 +1,35 @@
+  ///
+  /// \param[in] metaSkeletonStateSpace Statespace of the skeleton.
+  /// \param[in] metaskeleton Metaskeleton of the robot.
+  /// \param[in] nominalConfiguration Nominal Configuration. The current
```
Nit: `Nominal Configuration` -> `Nominal configuration`.

Can we also provide a `nullptr` default?

brianhou commented on src/distance/ConfigurationRanker.cpp:
```diff
@@ -0,0 +1,81 @@
+    throw std::invalid_argument("MetaSkeleton is nullptr.");
+
+  mDistanceMetric = createDistanceMetricFor(
+      std::dynamic_pointer_cast<statespace::CartesianProduct>(
```
Are both of these casts necessary? Does `createDistanceMetricFor` not work for `const statespace::CartesianProduct`?

aditya-vk replied:
The const-cast should not be necessary.

brianhou commented on src/distance/ConfigurationRanker.cpp:
```diff
@@ -0,0 +1,81 @@
+}
+
+//==============================================================================
+std::vector<statespace::CartesianProduct::State*>
```
It seems that we're returning raw pointers to the `CartesianProduct::State`s, rather than `CartesianProduct::ScopedState`s. What would happen if the input configurations are destructed before the ranking results are accessed?

brianhou commented on src/distance/JointAvoidanceConfigurationRanker.cpp:
```diff
@@ -0,0 +1,58 @@
+JointAvoidanceConfigurationRanker::JointAvoidanceConfigurationRanker(
+    ConstMetaSkeletonStateSpacePtr metaSkeletonStateSpace,
+    ConstMetaSkeletonPtr metaSkeleton)
+  : ConfigurationRanker(metaSkeletonStateSpace, metaSkeleton)
```
Should this `std::move` the `MetaSkeletonStateSpace` and `MetaSkeleton`?

brianhou commented:
This looks pretty good overall! Thanks for the changes.

@jslee02 would you mind looking over the `State` stuff to make sure we're using them correctly? Thanks!

aditya-vk commented:
Addressed the comments except for the `State`-handling. I'll wait for @jslee02 before modifications. I assumed that the way we'd be using the ranker we wouldn't be destroying the input states but that's a possibility in any case that needs taken care of.

brianhou commented:
I think it's reasonable to do something like `return ranker.rankConfigurations(states);`, which I think would lose the `states`.

jslee02 commented:
I'm not sure if I can review this PR soon (hopefully late this week?). Please go ahead if this is blocked by me. :sweat_smile:

jslee02 commented on tests/distance/test_JointAvoidanceConfigurationRanker.cpp:
```diff
@@ -0,0 +1,115 @@
+
+static constexpr double EPS = 1e-6;
+
+static BodyNode::Properties create_BodyNodeProperties(const std::string& _name)
```
Nit: No more leading underscore in parameter name!

jslee02 commented on src/distance/NominalConfigurationRanker.cpp:
```diff
@@ -0,0 +1,33 @@
+
+//==============================================================================
+double NominalConfigurationRanker::evaluateConfiguration(
+    statespace::dart::MetaSkeletonStateSpace::State* solution) const
```
Why `solution` is not const?

jslee02 commented on src/distance/NominalConfigurationRanker.cpp:
```diff
@@ -0,0 +1,33 @@
+namespace distance {
+
+using statespace::dart::ConstMetaSkeletonStateSpacePtr;
+using statespace::CartesianProduct;
```
Nit: This type alias isn't used

jslee02 commented on src/distance/JointAvoidanceConfigurationRanker.cpp:
```diff
@@ -0,0 +1,59 @@
+
+//==============================================================================
+double JointAvoidanceConfigurationRanker::evaluateConfiguration(
+    statespace::dart::MetaSkeletonStateSpace::State* solution) const
```
Maybe const?

jslee02 commented on src/distance/JointAvoidanceConfigurationRanker.cpp:
```diff
@@ -0,0 +1,59 @@
+        std::move(metaSkeletonStateSpace), std::move(metaSkeleton))
+{
+  auto lowerLimits = mMetaSkeleton->getPositionLowerLimits();
+  auto upperLimits = mMetaSkeleton->getPositionUpperLimits();
```
Nit: `const auto&` since `getPosition(Lower/Upper)Limits()` returns const reference.

jslee02 commented on src/distance/JointAvoidanceConfigurationRanker.cpp:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  mLowerLimitsState = mMetaSkeletonStateSpace->createState();
+  mUpperLimitsState = mMetaSkeletonStateSpace->createState();
```
`mUpperLimitsState` should be scoped state. Otherwise it will be destructed out of this function, and `mUpperLimitsState` will become dangling pointer.

brianhou commented on include/aikido/distance/ConfigurationRanker.hpp:
```diff
@@ -0,0 +1,54 @@
+  /// Destructor
+  virtual ~ConfigurationRanker() = default;
+
+  /// Ranks/Sorts the vector of configurations in increasing order of costs.
```
Nit: I think `sorts` is redundant here?

brianhou commented on include/aikido/distance/ConfigurationRanker.hpp:
```diff
@@ -0,0 +1,54 @@
+  virtual ~ConfigurationRanker() = default;
+
+  /// Ranks/Sorts the vector of configurations in increasing order of costs.
+  /// \param[in] configurations Vector of configurations to rank.
```
Nit: I believe the Doxygen convention for this is `\param[in,out]`.

brianhou commented on include/aikido/distance/JointAvoidanceConfigurationRanker.hpp:
```diff
@@ -0,0 +1,46 @@
+  std::vector<std::size_t> mUnboundedUpperLimitsIndices;
+
+  /// (Modified) State corresponding to the lower position limits.
+  /// The positions at the indices in mUnboundedLowerLimitsIndices are modified
```
I think we can change the descriptions for both lower/upper to clarify that the states are modified by `evaluateConfiguration` to match the input state before computing the distance from the lower limits.

brianhou commented on include/aikido/distance/ConfigurationRanker.hpp:
```diff
@@ -0,0 +1,54 @@
+
+namespace aikido {
+namespace distance {
+
```
Nit: document `ConfigurationRanker`.

brianhou commented on include/aikido/distance/JointAvoidanceConfigurationRanker.hpp:
```diff
@@ -0,0 +1,46 @@
+
+namespace aikido {
+namespace distance {
+
```
Nit: document `JointAvoidanceConfigurationRanker`.

I think we should mention here that only joints with limits contribute to the distance to the lower/upper limits.

brianhou commented on include/aikido/distance/NominalConfigurationRanker.hpp:
```diff
@@ -0,0 +1,37 @@
+
+namespace aikido {
+namespace distance {
+
```
Nit: document `NominalConfigurationRanker`.

brianhou commented on include/aikido/distance/NominalConfigurationRanker.hpp:
```diff
@@ -0,0 +1,37 @@
+  /// \param[in] metaSkeletonStateSpace Statespace of the skeleton.
+  /// \param[in] metaskeleton Metaskeleton of the robot.
+  /// \param[in] nominalConfiguration Nominal configuration. The current
+  /// configuration is considered if set to nullptr.
```
Nit: clarify that it's the current configuration of the `metaSkeleton`.

brianhou commented on include/aikido/distance/NominalConfigurationRanker.hpp:
```diff
@@ -0,0 +1,37 @@
+  /// Constructor
+  ///
+  /// \param[in] metaSkeletonStateSpace Statespace of the skeleton.
+  /// \param[in] metaskeleton Metaskeleton of the robot.
```
Nit: `metaskeleton` -> `metaSkeleton`.

brianhou commented on include/aikido/distance/JointAvoidanceConfigurationRanker.hpp:
```diff
@@ -0,0 +1,46 @@
+  /// Constructor
+  ///
+  /// \param[in] metaSkeletonStateSpace Statespace of the skeleton.
+  /// \param[in] metaskeleton Metaskeleton of the robot.
```
Nit: `metaskeleton` -> `metaSkeleton`.

brianhou commented on include/aikido/distance/ConfigurationRanker.hpp:
```diff
@@ -0,0 +1,54 @@
+  /// Constructor
+  ///
+  /// \param[in] metaSkeletonStateSpace Statespace of the skeleton.
+  /// \param[in] metaskeleton Metaskeleton of the robot.
```
Nit: `metaskeleton` -> `metaSkeleton`.

brianhou commented on src/distance/JointAvoidanceConfigurationRanker.cpp:
```diff
@@ -0,0 +1,58 @@
+  auto lowerLimits = mMetaSkeleton->getPositionLowerLimits();
+  auto upperLimits = mMetaSkeleton->getPositionUpperLimits();
+
+  for (std::size_t i = 0; i < mMetaSkeletonStateSpace->getDimension(); ++i)
```
Quick question: is there ever a case where there are lower limits but not upper limits (or vice versa)? If not, then rather than checking against `inf`s, maybe we can use `DegreeOfFreedom::hasPositionLimit`. I think it'd look something like this:
```cpp
auto properties = mMetaSkeletonStateSpace->getProperties();
for (std::size_t i = 0; i < properties.getNumDofs(); ++i)
{
  if (mMetaSkeleton->getDof(i)->hasPositionLimit())
    mUnboundedIndices.emplace_back(i);
}
```

jslee02 commented:
Hm, for some reason, [`test_NominalConfigurationRanker` fails on macOS](https://travis-ci.org/personalrobotics/aikido/jobs/404681752#L6011).
